### PR TITLE
[FLINK-17783][Table SQL/Ecosystem]Add array,map,row types support for orc row writer

### PR DESCRIPTION
**Pull request by wangwei1025**
## What is the purpose of the change

*Add array,map,row types support for orc row writer.*

## Brief change log

  - *Add array,map,row types support for RowDataVectorizer*

## Verifying this change

This change added tests and can be verified as follows:
  - *Added test that validates that RowDataVectorizer support array,map,row types and write correctly*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)(no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)(no)
  - The serializers: (yes / no / don't know)(no)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)(no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)(no)
  - The S3 file system connector: (yes / no / don't know)(no)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)(no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wangwei1025** [comment]
hi @JingsongLi , could you please help me review this PR? thanks.

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/RowDataVectorizer.java [hunk 1]
```diff
@@ -151,8 +160,114 @@ private static void setColumn(
...
+
+    /**
+     * Converting ArrayData to RowData for calling {@link RowDataVectorizer#setColumn(int,
+     * ColumnVector, LogicalType, RowData, int)} recursively with array.
+     *
+     * @param arrayData input ArrayData.
+     * @param arrayFieldType LogicalType of input ArrayData.
+     * @return RowData.
+     */
+    private static RowData convert(ArrayData arrayData, LogicalType arrayFieldType) {
```
Comment (JingsongLi) [inline flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/RowDataVectorizer.java:265]: This looks weird... But there is no better way...
I think you can update JIRA to explain why we need do this conversion.

**Review by JingsongLi — APPROVED** [review]
Thanks! Looks good to me!

**Comment by JingsongLi** [comment]
@lirui-apache Can u take a look too?

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/RowDataVectorizer.java (hunk 1 above)
Comment (wangwei1025) [reply flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/RowDataVectorizer.java:265]: OK, I append a comment to explain the RowDataVectorizer.convert function.

**Review by lirui-apache — COMMENTED** [review]
Thanks @wangwei1025 for the PR. LGTM except for a minor comment.

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/RowDataVectorizer.java [hunk 2]
```diff
@@ -151,8 +160,114 @@ private static void setColumn(
...
+        listColumnVector.offsets[rowId] = listColumnVector.childCount;
+        listColumnVector.childCount += listColumnVector.lengths[rowId];
+        listColumnVector.child.ensureSize(
+                listColumnVector.childCount, listColumnVector.offsets[rowId] != 0);
+        for (int i = 0; i < arrayData.size(); i++) {
+            setColumn(
+                    (int) listColumnVector.offsets[rowId] + i,
+                    listColumnVector.child,
+                    arrayType.getElementType(),
+                    convert(arrayData, arrayType.getElementType()),
```
[the remark this answers is not in the document]
Comment (wangwei1025) [reply flink-formats/flink-orc/src/main/java/org/apache/flink/orc/vector/RowDataVectorizer.java]: Yes, it's uncessary. This happend in setColumn(int,MapColumnVector,LogicalType,RowData,int)  too. I fix them in this latest commit [hotfix]move convert out of loop

**Review by lirui-apache — APPROVED** [review]
Thanks for updating. +1 pending test

**Pull request closed without merging**